pkg/didcomm/protocol/didexchange/service_test.go

```diff
@@ -122,182 +122,228 @@ func TestSendResponse(t *testing.T) {
-			Payload: request}
-		err = s.Handle(msg)
-		require.NoError(t, err)
+		err = s.Handle(dispatcher.DIDCommMsg{Type: connectionResponse, Payload: response})
+		require.Error(t, err)
 	})
-	t.Run("good state transition without an invitation", func(t *testing.T) {
-		request, err := json.Marshal(
-			&Request{
-				Type:       "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/didexchange/1.0/request",
-				ID:         randomString(),
-				Label:      "test",
-				Connection: &Connection{DID: "did:example:123"},
+	t.Run("must not start with ACK msg", func(t *testing.T) {
+		s := &Service{outboundTransport: newMockOutboundTransport(), store: newMockStore()}
+		ack, err := json.Marshal(
+			&Ack{
+				Type: connectionAck,
+				ID:   randomString(),
 			},
 		)
 		require.NoError(t, err)
-		msg := dispatcher.DIDCommMsg{
-			Type:    "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/didexchange/1.0/request",
-			Payload: request}
-		err = s.Handle(msg)
-		require.NoError(t, err)
+		err = s.Handle(dispatcher.DIDCommMsg{Type: connectionAck, Payload: ack})
+		require.Error(t, err)
 	})
-
-	t.Run("bad state transition", func(t *testing.T) {
+	t.Run("must not transition to same state", func(t *testing.T) {
+		s := &Service{outboundTransport: newMockOutboundTransport(), store: newMockStore()}
 		thid := randomString()
-		invitation, err := json.Marshal(
-			&Invitation{
-				Type:  "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/didexchange/1.0/invitation",
+		request, err := json.Marshal(
+			&Request{
+				Type:  connectionRequest,
 				ID:    thid,
-				Label: "Alice",
-				DID:   "did:sov:QmWbsNYhMrjHiqZDTUTEJs",
-			})
+				Label: "test",
+			},
+		)
 		require.NoError(t, err)
-		msg := dispatcher.DIDCommMsg{
-			Type:    "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/didexchange/1.0/invitation",
-			Payload: invitation}
-		err = s.Handle(msg)
+		err = s.Handle(dispatcher.DIDCommMsg{Type: connectionRequest, Outbound: false, Payload: request})
 		require.NoError(t, err)
-
+		// state machine has automatically transitioned to responded state
+		actual, err := s.currentState(thid)
+		require.NoError(t, err)
+		require.Equal(t, (&responded{}).Name(), actual.Name())
+		// therefore cannot transition Responded state
 		response, err := json.Marshal(
 			&Response{
-				Type:   "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/didexchange/1.0/response",
+				Type:   connectionResponse,
 				ID:     randomString(),
 				Thread: &decorator.Thread{ID: thid},
-				ConnectionSignature: &ConnectionSignature{
-					Type: "did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/signature/1.0/ed25519Sha512_single",
+			},
+		)
+		require.NoError(t, err)
+		err = s.Handle(dispatcher.DIDCommMsg{Type: connectionResponse, Outbound: false, Payload: response})
+		require.Error(t, err)
+	})
+	t.Run("error when updating store on first state transition", func(t *testing.T) {
+		s := &Service{
+			outboundTransport: newMockOutboundTransport(),
+			store: &mockStore{
+				get: func(string) ([]byte, error) {
+					return nil, errors.New("not found")
```

Comment:
we should probably use a preset Not Found error, like this one: https://github.com/hyperledger/aries-framework-go/pull/236/files#diff-f761c91e753eba1c42a68076df06ce12R97

Reply:
#195 was opened for that